components/content/content-editor/controls/block/controls.ts

```diff
@@ -0,0 +1,203 @@
+import { Node, NodeType } from 'prosemirror-model';
+import { Selection } from 'prosemirror-state';
+import { EditorView } from 'prosemirror-view';
+import Vue from 'vue';
+import { Component, Prop, Watch } from 'vue-property-decorator';
+import { AppTooltip } from '../../../../../components/tooltip/tooltip';
+import { Growls } from '../../../../growls/growls.service';
+import { Screen } from '../../../../screen/screen-service';
+import { ContextCapabilities } from '../../../content-context';
+import { ContentEditorService } from '../../content-editor.service';
+import { ContentTableService } from '../../content-table.service';
+
+@Component({
+	components: {},
+	directives: {
+		AppTooltip,
+	},
+})
+export default class AppContentEditorBlockControls extends Vue {
+	@Prop(Object)
+	view!: EditorView;
+	@Prop(Number)
+	stateCounter!: number;
+	@Prop(Object)
+	capabilities!: ContextCapabilities;
+	@Prop(Boolean)
+	collapsed!: boolean;
+
+	visible = false;
+	top = '0px';
+	left = '0px';
+
+	readonly Screen = Screen;
+
+	$refs!: {
+		container: HTMLElement;
+	};
+
+	mounted() {
+		this.update();
+	}
+
+	@Watch('stateCounter')
+	private update() {
+		if (this.view instanceof EditorView) {
+			const state = this.view.state;
+			const node = ContentEditorService.getSelectedNode(this.view.state);
+
+			if (node !== null && node.type.name === 'paragraph' && state.selection.empty) {
+				this.visible = true;
+
+				const start = this.view.coordsAtPos(state.selection.from);
+
+				// Offset the controls if there's a heading node.
+				const headingNode = this.testIsInHeading(node);
+				if (headingNode instanceof Node) {
+					if (headingNode.attrs.level === 1) {
+						start.top += 8;
+					} else {
+						start.top += 6;
+					}
+				}
+
+				const box = this.$refs.container.offsetParent.getBoundingClientRect();
+				this.top = start.top - box.top - 10 + 'px';
+				this.left = '-32px';
+
+				return;
+			}
+		}
+		this.visible = false;
+		this.setCollapsed(true);
+	}
+
+	testIsInHeading(node: Node) {
+		if (!this.capabilities.heading) {
+			return false;
+		}
+		return ContentEditorService.isContainedInNode(
+			this.view.state,
+			node,
+			this.view.state.schema.nodes.heading
+		);
+	}
+
+	private setCollapsed(value: boolean) {
+		this.$emit('collapsedChanged', value);
+	}
+
+	onClickExpand() {
+		this.setCollapsed(!this.collapsed);
+	}
+
+	private insertNewNode(newNode: Node) {
+		const tr = this.view.state.tr;
+		tr.replaceWith(
+			this.view.state.selection.from - 1,
```

Comment:
Replaces the empty paragraph.

Reply:
The block controls only show when you are INSIDE an empty paragraph (that we always ensure we have). The -1 and +1 makes it replace the paragraph itself, instead of just the contents within it to the block level element we are using.